### PR TITLE
PSA Integration: Add support for client-side PSA-based ECDHE

### PR DESCRIPTION
**Pull request by hanno-arm**
__Summary:__ This PR adds support for client-side PSA-based pure ECDHE (that is, excluding ECDHE-PSK).

Please see the commit messages for more information.

__Points of discussion:__ So far, the PR cuts corners on efficiency and code sharing in some respects: 
* The peer's public key is stored within the `mbedtls_ssl_handshake_params` structure between the receipt of the `ServerKeyExchange` message and the derivation of the key material in `mbedtls_ssl_derive_keys()` at the end of the handshake. In principle, the premaster secret could be generated straight away when reading the `ServerKeyExchange` message.
* The legacy ECDH context is still unconditionally contained as a field of `mbedtls_ssl_handshake_params`. We might consider adding a bunch of `#define`'s around this to ensure that in builds which only use PSA-based client-side TLS 1.2 with pure-ECDHE, it is omitted from the structure.
* ~~The ECDH shared secret is extracted into a key slot which is then used in the PSA-based TLS 1.2 PRF; this could be unified with the default, so far non-PSA based PRF once @AndrzejKurek rewrites the latter to use PSA.~~

__Dependencies on PSA Crypto__
* The PSA Crypto API suggests the combined use of key agreement + KDF. With the current API, though, this cannot yet be done in our use case, as we need to pass non-NULL values for salt+label to the TLS 1.2 PRF and the PSA key agreement API doesn't support that yet when used in combination with a key agreement scheme.
* The code depends on the EC public key format used by PSA; ~~in the version used by this PR, PSA uses the `SubjectPublicKeyInfo` structure, but that will change to the `ECPoint` structure used in the `ServerKeyExchange` message at some point. The PR will simplify considerably with this forthcoming change.~~ The PR has been adapted to the new simplified format introduced in https://github.com/ARMmbed/mbed-crypto/pull/13.

__Gatekeeping note:__
This PR is based on #2382 which should be merged first.
[added by mpg] actually now this PR should be rebased on development once #2395 has been merged.

### CODE REVIEW

**Comment by hanno-arm** [comment]
There's a bug in the PR for TLS 1.1 and 1.0. Demoting to 'needs work'.

**Comment by hanno-arm** [comment]
Simplified the PR and hopefully fixed the issue with TLS 1.0 and 1.1. @AndrzejKurek @mpg Please review.

File: include/mbedtls/ssl_internal.h [hunk 1]
```diff
@@ -270,7 +270,15 @@ struct mbedtls_ssl_handshake_params
...
 #if defined(MBEDTLS_ECDH_C)
     mbedtls_ecdh_context ecdh_ctx;              /*!<  ECDH key exchange       */
+
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    psa_ecc_curve_t ecdh_psa_curve;
+    psa_key_slot_t ecdh_psa_privkey;
+    unsigned char ecdh_psa_peerkey[256];
+    size_t ecdh_psa_peerkey_len;
+#endif /* MBEDTLS_USE_PSA_CRYPTO */
 #endif
```
Comment (AndrzejKurek) [inline include/mbedtls/ssl_internal.h]: Please add a 'MBEDTLS_ECDH_C' comment here, just for clarity.

File: library/Makefile [hunk 2]
```diff
@@ -88,10 +88,11 @@ OBJS_CRYPTO=	aes.o		aesni.o		arc4.o		\
 		pk.o		pk_wrap.o	pkcs12.o	\
 		pkcs5.o		pkparse.o	pkwrite.o	\
 		platform.o	platform_util.o	poly1305.o	\
-		ripemd160.o	rsa_internal.o	rsa.o  		\
-		sha1.o		sha256.o	sha512.o	\
-		threading.o	timing.o	version.o	\
-		version_features.o		xtea.o
+                ripemd160.o	rsa_internal.o	\
```
Comment (AndrzejKurek) [inline library/Makefile]: Why did you change the formatting here?

File: library/ssl_cli.c [hunk 3]
```diff
@@ -2976,6 +3056,113 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
...
+          defined(MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED) )
+    if( ciphersuite_info->key_exchange == MBEDTLS_KEY_EXCHANGE_ECDHE_RSA ||
+        ciphersuite_info->key_exchange == MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA )
+    {
+        psa_status_t status;
+        psa_key_policy_t policy;
+
+        mbedtls_ssl_handshake_params *handshake = ssl->handshake;
+
+        unsigned char own_pubkey[256];
```
Comment (AndrzejKurek) [inline library/ssl_cli.c]: Shouldn't this come from a constant defined somewhere?

File: library/ssl_cli.c [hunk 4]
```diff
@@ -2976,6 +3056,113 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
...
+        size_t own_pubkey_len;
+        unsigned char *own_pubkey_ecpoint;
+        size_t own_pubkey_ecpoint_len;
+
+        psa_crypto_generator_t generator = PSA_CRYPTO_GENERATOR_INIT;
+
+        i = 4;
```
Comment (AndrzejKurek) [inline library/ssl_cli.c]: Please comment on what does this mean.

File: library/ssl_cli.c [hunk 5]
```diff
@@ -2976,6 +3056,113 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
...
+         * of the ECDH secret and the application of the TLS 1.2 PRF. */
+        psa_key_policy_init( &policy );
+        psa_key_policy_set_usage( &policy,
+                               PSA_KEY_USAGE_DERIVE,
+                               PSA_ALG_ECDH( PSA_ALG_SELECT_RAW ) );
+        status = psa_set_key_policy( handshake->ecdh_psa_privkey, &policy );
+        if( status != PSA_SUCCESS )
+            return( MBEDTLS_ERR_SSL_HW_ACCEL_FAILED );
+
+
```
Comment (AndrzejKurek) [inline library/ssl_cli.c]: Double empty lines.

File: library/ssl_cli.c [hunk 6]
```diff
@@ -2976,6 +3056,113 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
...
+        /* Generate ECDH private key. */
+        status = psa_generate_key( handshake->ecdh_psa_privkey,
+                          PSA_KEY_TYPE_ECC_KEYPAIR( handshake->ecdh_psa_curve ),
+                          MBEDTLS_PSA_ECC_KEY_BITS_OF_CURVE( handshake->ecdh_psa_curve ),
+                          NULL, 0 );
+        if( status != PSA_SUCCESS )
+            return( MBEDTLS_ERR_SSL_HW_ACCEL_FAILED );
+
+        /* Export the public part fo the ECDH private key from PSA
```
Comment (AndrzejKurek) [inline library/ssl_cli.c]: Typo - "fo" -> "of"

Comment (hanno-arm) [reply library/ssl_cli.c]: Fixing.

File: library/ssl_cli.c (hunk 5 above)
Comment (hanno-arm) [reply library/ssl_cli.c]: Fixing.

File: library/ssl_cli.c (hunk 4 above)
Comment (hanno-arm) [reply library/ssl_cli.c]: That's pre-existing, but I've renamed the variables `i,n` throughout to make their meaning clearer.

File: library/Makefile (hunk 2 above)
Comment (hanno-arm) [reply library/Makefile]: I initially added `psa_util.o` to the crypto object file list, which was wrong, and when I moved it to the TLS object file list, the reformatting of the crypto list remained. I'll revert it.

File: include/mbedtls/ssl_internal.h (hunk 1 above)
Comment (hanno-arm) [reply include/mbedtls/ssl_internal.h]: Yes.

File: library/ssl_cli.c (hunk 3 above)
Comment (hanno-arm) [reply library/ssl_cli.c]: Fixing.

**Comment by hanno-arm** [comment]
@AndrzejKurek Thank you for your review! I believe to have addressed your comments. Could you please re-review?

File: include/mbedtls/ssl_internal.h [hunk 7]
```diff
@@ -270,7 +270,15 @@ struct mbedtls_ssl_handshake_params
...
 #if defined(MBEDTLS_ECDH_C)
     mbedtls_ecdh_context ecdh_ctx;              /*!<  ECDH key exchange       */
-#endif
+
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    psa_ecc_curve_t ecdh_psa_curve;
+    psa_key_slot_t ecdh_psa_privkey;
+    unsigned char ecdh_psa_peerkey[256];
```
Comment (AndrzejKurek) [inline include/mbedtls/ssl_internal.h]: Should this also be MBEDTLS_PSA_MAX_EC_PUBKEY_LENGTH ?

Comment (hanno-arm) [reply include/mbedtls/ssl_internal.h]: Yes

File: include/mbedtls/psa_util.h [hunk 8]
```diff
@@ -259,6 +279,121 @@ static inline int mbedtls_psa_err_translate_pk( psa_status_t status )
...
+#endif /* MBEDTLS_ECP_DP_BP512R1_ENABLED */
+    }
+
+    return( -1 );
+}
+
+/* This function takes a buffer holding an EC public key
+ * exported through psa_export_public_key(), and converts
+ * it into an ECPoint structure to be put into a ClientKeyExchange
+ * message in an ECDHE exchange).
```
Comment (AndrzejKurek) [inline include/mbedtls/psa_util.h]: Stray bracket?

File: library/psa_util.c [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+     *      algorithm          AlgorithmIdentifier,
+     *      subjectPublicKey   BIT STRING  }
+     * AlgorithmIdentifier  ::=  SEQUENCE  {
+     *      algorithm          OBJECT IDENTIFIER,
+     *      parameters         ANY DEFINED BY algorithm OPTIONAL  }
+     * ```
+     *
+     * In our context, these fields have the following values:
+     * - `subjectPublicKey` is an `ECPoint` structure
+     *   and it passed in the source buffer src.
```
Comment (AndrzejKurek) [inline library/psa_util.c]: it is passed?

File: library/ssl_cli.c [hunk 10]
```diff
@@ -2976,6 +3058,113 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
...
+        /* The master secret is obtained from the shared ECDH secret by
+         * applying the TLS 1.2 PRF with a specific salt and label. While
+         * the PSA Crypto API encourages combining key agreement schemes
+         * such as ECDH with fixed KDFs such as TLS 1.2 PRF, it does not
+         * yet support the provisioning of salt + label to the KDF.
+         * For the time being, we therefore need to split the computation
+         * of the ECDH secret and the application of the TLS 1.2 PRF. */
+        psa_key_policy_init( &policy );
+        psa_key_policy_set_usage( &policy,
+                               PSA_KEY_USAGE_DERIVE,
```
Comment (AndrzejKurek) [inline library/ssl_cli.c]: I'm not sure if this line and the one below is indented properly.

File: library/ssl_cli.c [hunk 11]
```diff
@@ -2976,6 +3058,113 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
...
+                          NULL, 0 );
+        if( status != PSA_SUCCESS )
+            return( MBEDTLS_ERR_SSL_HW_ACCEL_FAILED );
+
+        /* Export the public part of the ECDH private key from PSA
+         * and convert it to ECPoint format used in ClientKeyExchange. */
+        status = psa_export_public_key( handshake->ecdh_psa_privkey,
+                                        own_pubkey, sizeof( own_pubkey ),
+                                        &own_pubkey_len );
+        if( status != PSA_SUCCESS )
```
Comment (AndrzejKurek) [inline library/ssl_cli.c]: Shouldn't we deal with the `handshake->ecdh_psa_privkey` key upon encountering failures in this function? Or is it dealt with the handshake context freeing function if it's not owned?

File: library/ssl_cli.c [hunk 12]
```diff
@@ -3062,26 +3251,28 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
             return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
         }
 
-        i = 4;
-        n = ssl->conf->psk_identity_len;
+        header_len = 4;
+        content_len = ssl->conf->psk_identity_len;
 
-        if( i + 2 + n > MBEDTLS_SSL_OUT_CONTENT_LEN )
+        if( header_len + 2 + content_len > MBEDTLS_SSL_OUT_CONTENT_LEN )
```
Comment (AndrzejKurek) [inline library/ssl_cli.c]: This "2" constant still bothers me, but it's preexisting.

File: library/ssl_cli.c [hunk 13]
```diff
@@ -3094,7 +3285,8 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
                 return( MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE );
 #endif /* MBEDTLS_USE_PSA_CRYPTO */
 
-            if( ( ret = ssl_write_encrypted_pms( ssl, i, &n, 2 ) ) != 0 )
+            if( ( ret = ssl_write_encrypted_pms( ssl, header_len,
+                                                 &content_len, 2 ) ) != 0 )
```
Comment (AndrzejKurek) [inline library/ssl_cli.c]: Same as above about this "2".

File: library/ssl_cli.c [hunk 14]
```diff
@@ -3111,21 +3303,22 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
             /*
              * ClientDiffieHellmanPublic public (DHM send G^X mod P)
              */
-            n = ssl->handshake->dhm_ctx.len;
+            content_len = ssl->handshake->dhm_ctx.len;
 
-            if( i + 2 + n > MBEDTLS_SSL_OUT_CONTENT_LEN )
+            if( header_len + 2 + content_len >
```
Comment (AndrzejKurek) [inline library/ssl_cli.c]: Same as above about this "2".

File: include/mbedtls/psa_util.h (hunk 8 above)
Comment (hanno-arm) [reply include/mbedtls/psa_util.h]: Fixed.

File: library/psa_util.c (hunk 9 above)
Comment (hanno-arm) [reply library/psa_util.c]: Fixed.

File: library/ssl_cli.c (hunk 10 above)
Comment (hanno-arm) [reply library/ssl_cli.c]: That was to keep them below 80 characters when the TLS 1.2 PRF identifier was used in place of `PSA_ALG_SELECT_RAW`. Fixed.

File: library/ssl_cli.c (hunk 11 above)
Comment (hanno-arm) [reply library/ssl_cli.c]: Yes, the key destruction is done in `ssl_handshake_free()`. However, I think it's safer style to destroy the key as soon as its no longer needed, so I also added a call to `psa_destroy_key()` at the end of this function.

File: library/ssl_cli.c (hunk 12 above)
Comment (hanno-arm) [reply library/ssl_cli.c]: I agree that the code isn't very readable, but that's pre-existing and I prefer not to change it in this PR.

File: library/ssl_cli.c (hunk 13 above)
Comment (hanno-arm) [reply library/ssl_cli.c]: Same here.

File: library/ssl_cli.c (hunk 14 above)
Comment (hanno-arm) [reply library/ssl_cli.c]: Same here.

**Comment by hanno-arm** [comment]
@AndrzejKurek Thanks for another round of review. I think I have addressed your comments. Could you please re-review?

**Comment by mpg** [comment]
@hanno-arm As discussed by email, let's wait until we have the public key format for finalizing this. I'm labeling as "needs: preceding PR" and "needs: work" to remind us of that, even though I don't think that's work you can actually start now.

**Comment by hanno-arm** [comment]
@mpg @AndrzejKurek This PR will significantly simplify (specifically: the wrappers `mbedtls_psa_tls_psa_ec_to_ecpoint()` and `mbedtls_psa_tls_ecpoint_to_psa_ec()`, which currently perform a non-trivial transformation of data, will essentially become no-ops) once the new public key format has been merged into PSA.

**Comment by hanno-arm** [comment]
@AndrzejKurek @mpg This PR has now been rebased on top of #2382 and simplified accordingly.

Please review.

**Comment by mpg** [comment]
@hanno-arm The CI failed on `check-generated-files.sh`, can you investigate that? Thanks!

**Comment by hanno-arm** [comment]
@mpg Fixed.

@mpg @AndrzejKurek Please re-review.

**Comment by mpg** [comment]
(Changed the base twice in order to refresh github's cache.)

File: tests/ssl-opt.sh [hunk 15]
```diff
@@ -769,6 +769,17 @@ run_test_psa() {
                 -C "error"
 }
 
+run_test_psa_force_curve() {
+    requires_config_enabled MBEDTLS_USE_PSA_CRYPTO
+    run_test    "PSA - ECDH with $1" \
+                "$P_SRV debug_level=4 force_version=tls1_2" \
+                "$P_CLI debug_level=4 force_version=tls1_2 force_ciphersuite=TLS-ECDHE-RSA-WITH-AES-128-GCM-SHA256 curves=$1" \
+                0 \
+                -s "Protocol is TLSv1.2" \
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply tests/ssl-opt.sh]: Yes.

File: include/mbedtls/psa_util.h [hunk 16]
```diff
@@ -157,6 +157,8 @@ static inline psa_algorithm_t mbedtls_psa_translate_md( mbedtls_md_type_t md_alg
 
 /* Translations for ECC. */
 
+#define MBEDTLS_PSA_MAX_EC_PUBKEY_LENGTH 256
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/psa_util.h]: It was meant to be an overapproximation of the maximum length of EC public keys. I've refined the definition in 135baef1bd45c78ca73d9861eb7155449591acd2.

File: include/mbedtls/psa_util.h [hunk 17]
```diff
@@ -235,7 +235,85 @@ static inline int mbedtls_psa_get_ecc_oid_from_id(
      return( -1 );
 }
 
-#define MBEDTLS_PSA_MAX_EC_PUBKEY_LENGTH 256
+#define MBEDTLS_PSA_MAX_EC_PUBKEY_LENGTH 1
```
Comment (mpg) [inline include/mbedtls/psa_util.h]: Thanks for fixing this. We should probably ask if this should be provided by PSA, as I think they have a policy to provide macros for the maxiumx size of each possible output buffers to their functions, which I think could cover this (or close enough that we don't have to redo everything from scratch here.)

(This should not block this PR, just making a note for the future.)

**Comment by mpg** [comment]
Removing "needs: preceding PR" as #2395 has been merged now.

**Comment by sbutcher-arm** [comment]
Reviewed, approved and passed CI, therefore labelling as 'ready for merge'.

**Comment by hanno-arm** [comment]
The PR broke the build because it didn't consider that restartable ECC doesn't work with PSA. We must catch this in `check_config.h`, and no test in `all.sh` should set them simultaneously. The latest commits address this.

**Comment by Patater** [comment]
@mpg approved up until https://github.com/ARMmbed/mbedtls/commit/1ce51e4dc3b459d2d27ed563b030055294468859
@k-stachowiak approved up until https://github.com/ARMmbed/mbedtls/commit/1ce51e4dc3b459d2d27ed563b030055294468859 and then additionally approved https://github.com/ARMmbed/mbedtls/commit/1ce51e4dc3b459d2d27ed563b030055294468859 and https://github.com/ARMmbed/mbedtls/commit/241b524964d70b829c43c7774a1bc0e93dbe9354
